funcs/func_groupcount.c

```diff
@@ -51,37 +56,200 @@
+			</example>
+		</description>
+		<see-also>
+			<ref type="function">GROUP</ref>
+			<ref type="function">GROUP_MATCH_LIST_NEXT</ref>
+			<ref type="function">GROUP_CHANNEL_LIST</ref>
+		</see-also>
+	</function>
+	<function name="GROUP_MATCH_LIST_NEXT" language="en_US">
+		<synopsis>
+			Get a match from the GROUP_MATCH_LIST_START search currently loaded
+		</synopsis>
+		<description>
+			<para>Once GROUP_MATCH_LIST_START is executed with a search, then each call to GROUP_MATCH_LIST_NEXT()
+			will return a matched group.  This search will look at all known groups and filter by the regex provided.</para>
+			<para>The return value is the number of groups found. Calling GROUP_MATCH_LIST_NEXT() more times than there
+			are matches will result in an empty value being returned.</para>
+			<para>Note: the search is executed and stored at the time of calling this function.
+			If GROUP() assignments change during successive calls to the _NEXT function, you will need to call _START again
+			to see the new groups. An empty string is returned to indicate the end of the list.</para>
+			<para>Uses standard regular expression matching (see regex(7)).</para>
+			<example title="Start a search and return the number of groups found">Set(group_found=${GROUP_MATCH_LIST_NEXT()})</example>
+		</description>
+		<see-also>
+			<ref type="function">GROUP</ref>
+			<ref type="function">GROUP_MATCH_LIST_START</ref>
+		</see-also>
+	</function>
+	<function name="GROUP_CHANNEL_LIST" language="en_US">
+		<synopsis>
+			Start a find of groups by regular expression
+		</synopsis>
+		<syntax argsep="@">
+			<parameter name="group_regex"/>
+			<parameter name="category_regex"/>
+		</syntax>
+		<description>
+			<para>Once GROUP_MATCH_LIST_START is executed with a search, then each call to GROUP_MATCH_LIST_NEXT()
+			will return a matched group.  This search will look at all known groups and filter by the regex provided.</para>
+			<para>It is possible to only specify @category_regex, and it is also possible to pass in nothing for the search.</para>
+			<para>If only the category_regex is specified, then it will only return group@category that matches this category search.</para>
+			<para>If the search regex is entirely empty, all groups will be returned.</para>
+			<para>The return value is the number of groups found. Calling GROUP_MATCH_LIST_NEXT() more times than there
+			are matches will result in an empty value being returned.</para>
+			<para>Note: the search is executed and stored at the time of calling this function.
+			If GROUP() assignments change during successive calls to the _NEXT function, you will need to call _START again
+			to see the new groups. An empty string is returned to indicate the end of the list.</para>
+			<para>Uses standard regular expression matching (see regex(7)).</para>
+			<example>Find groups containing 'foo': ${GROUP_MATCH_LIST_START(.*foo.*)}</example>
```

Comment:
I'm pretty sure I raised this issue in the Gerrit review but this has not been fixed. These are not valid examples and this documentation is malformed. There should be titles, e.g.

```
<example title="foo bar">
some example
</example>
```

```
PHP Fatal error:  Uncaught ErrorException: Undefined array key "title" in /usr/src/asterisk-20.4.0/astdocgen.php:431
Stack trace:
#0 /usr/src/asterisk-20.4.0/astdocgen.php(431): exception_error_handler()
#1 {main}
  thrown in /usr/src/asterisk-20.4.0/astdocgen.php on line 431
```

Reply:
Thanks... this is getting rewritten anyway as mentioned previously.  The new version of this will return a hash.

The biggest issue with the current return types from most functions is they are not data-agnostic.  For example if you return a comma delimited list and your group names have commas, it's not possible to parse it properly.  The problem is that in dialplan, data arguments are not isolated from each other like in a traditional language.  So you would have this sort of issue no matter what delimiter you used (if the delimiter was contained within the data)

Reply:
Okay, fair enough, I think the only problems were in this part of the documentation. I just manually deleted the `<example>` lines in the core-doc file so I could build my docs.

I don't use commas or other possible delimiters in my group names so I guess I haven't seen this kind of issue... usually they are something pretty simple like [unique identifier]@[some property of relevance]. At this point most of my group usage is actually in C, not in dialplan, typically GROUP_COUNT or GROUP_VARs now :)

I think you should put out a feeler on the -dev list about this, as I think there should be a more generic mechanism, if at all, that is not specific to groups. To me it feels like if somebody needs this type of thing, it would be best done in a C module where you can directly iterate over what you want in a structured programming language with a proper iterator, rather than trying to deal with parsing to begin with.